### PR TITLE
Experimental conformance promotion to standard

### LINKED ISSUE #2851: Promote the experimental conformance test suite to standard
**What would you like to be added**:

The experimental conformance test suite is going to be promoted to standard. In order to do so, we need to:
- remove all the `experimental` keywords used in the suite code
- replace the old standard suite with the new one

### PR DESCRIPTION
**Pull request by mlavacca**
**What type of PR is this?**

/area conformance

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #2851

**Does this PR introduce a user-facing change?**:

```release-note
The Experimental Conformance test suite has been promoted to standard. Furthermore, the `ConformanceReport` API has been promoted to `v1`.
```

### CODE REVIEW

File: conformance/utils/suite/profiles.go
Comment (dprotaso) [inline conformance/utils/suite/profiles.go:71]: Adding Gateway Extended Features to HTTP profile seems off

I could imagine there being an Extended feature that has nothing to do with HTTPRoute

File: conformance/utils/flags/flags.go
Comment (mlavacca) [reply conformance/utils/flags/flags.go:28]: It's used in row 47

File: conformance/utils/suite/conformance-test.go
Comment (mlavacca) [reply conformance/utils/suite/conformance-test.go:27]: I totally agree, but `conformance_test.go` does not fit this file, as it does not contain tests. The alternative is `test.go` (with a `test_test.go` file that contains tests for the `test` file). Do you think `test` is better here?

File: conformance/conformance_test.go
Comment (mlavacca) [reply conformance/conformance_test.go]: Yeah it was not really intended, it's some copy-paste side effect. Reverted.

File: conformance/utils/suite/profiles.go
Comment (mlavacca) [reply conformance/utils/suite/profiles.go:71]: The current Gateway API extended features are `GatewayPort8080` and `GatewayStaticAddresses`. I think both are totally relevant to the HTTP profile and should be included here.

File: conformance/utils/suite/conformance-test.go
Comment (dprotaso) [reply conformance/utils/suite/conformance-test.go:27]: Just call it `conformance.go` or `conformancetest.go`

File: conformance/utils/suite/profiles.go
Comment (dprotaso) [reply conformance/utils/suite/profiles.go:71]: Woud gateway infrastructure be something that we want in the HTTP profile?

https://github.com/kubernetes-sigs/gateway-api/pull/2845

File: conformance/utils/flags/flags.go
Comment (mlavacca) [reply conformance/utils/flags/flags.go:28]: It's used in the cli, here https://github.com/mlavacca/gateway-api/blob/4ea1e107e62eb89c3581392d7dea42f29ac4eda3/conformance/conformance_test.go#L83

File: conformance/utils/suite/profiles.go
Comment (mlavacca) [reply conformance/utils/suite/profiles.go:71]: I think so, WDYT?

File: conformance/utils/flags/flags.go
Comment (dprotaso) [reply conformance/utils/flags/flags.go:28]: ah i missed that - thanks

File: conformance/utils/suite/conformance-test.go
Comment (mlavacca) [reply conformance/utils/suite/conformance-test.go]: here it is: [5bc530f](https://github.com/kubernetes-sigs/gateway-api/pull/2868/commits/5bc530f61014ab65ba18e48785243793f9aeebd5)

File: conformance/apis/v1beta1/doc.go
Comment (robscott) [inline conformance/apis/v1beta1/doc.go:23]: Nit: In our updated versioning policy, standard means "v1".
```suggestion
package v1
```

File: conformance/utils/flags/flags.go
Comment (robscott) [inline conformance/utils/flags/flags.go:45]: Nit: I think "to issue conformance to" is a bit redundant here.

File: conformance/utils/suite/conformance.go
Comment (robscott) [inline conformance/utils/suite/conformance.go:2]: I can't tell if this is a new file or if Git's just not doing a great job at tracking movement - if it is new, it should be 2024 here.

File: conformance/utils/suite/conformance_test.go
Comment (robscott) [inline conformance/utils/suite/conformance_test.go:2]: Same as above, not sure if this is new.

File: conformance/utils/suite/suite.go
Comment (robscott) [inline conformance/utils/suite/suite.go:97]: Can you remind me why we need this? Assuming we do, it would be useful to include the rationale in the comment for future reviewers.

File: conformance/utils/suite/suite.go
Comment (robscott) [inline conformance/utils/suite/suite.go:159]: Why do we need this? A comment would be helpful

File: conformance/apis/v1beta1/profilereport.go
Comment (robscott) [inline conformance/apis/v1beta1/profilereport.go:20]: Any harm in moving away from the "Conformance Profiles" term and just using "Conformance Report" here? If I'm remembering correctly, we've basically just said that you support a resource or not, but have mostly moved away from profiles that group resources together?

**Review by robscott — COMMENTED** [review]
Thanks @mlavacca!

File: conformance/apis/v1beta1/doc.go
Comment (mlavacca) [reply conformance/apis/v1beta1/doc.go:23]: oh you are right 👍

File: conformance/utils/suite/conformance.go
Comment (mlavacca) [reply conformance/utils/suite/conformance.go:2]: It's old code, just moved it to another file

File: conformance/apis/v1beta1/profilereport.go
Comment (mlavacca) [reply conformance/apis/v1beta1/profilereport.go:20]: I'm not sure I'm getting your point here, but `ConformanceProfiles` in the report is still a slice containing multiple profile results: 

https://github.com/kubernetes-sigs/gateway-api/blob/7c672a75aaf5e66b05f9df5f2c7b3ce2d8e03d9f/conformance/apis/v1alpha1/conformancereport.go#L44-L46

File: conformance/utils/suite/conformance_test.go
Comment (mlavacca) [reply conformance/utils/suite/conformance_test.go:2]: It's old code, just moved to another file

File: conformance/utils/suite/conformance.go
Comment (mlavacca) [reply conformance/utils/suite/conformance.go:2]: updated

File: conformance/utils/suite/conformance_test.go
Comment (mlavacca) [reply conformance/utils/suite/conformance_test.go:2]: updated

File: conformance/utils/suite/suite.go
Comment (mlavacca) [reply conformance/utils/suite/suite.go:97]: I've updated the comment:
```go
	// running indicates whether the test suite is currently running.
	// Throught this flag we prevent a Run() execution to happen in case
	// another Run() is ongoing.
	running bool
```

File: conformance/utils/suite/suite.go
Comment (mlavacca) [reply conformance/utils/suite/suite.go:159]: I've added a comment:
```go
	// undefinedKeyword is set in the ConformanceReport "GatewayAPIVersion"
	// and "GatewayAPIChannel" fields in case it's not possible to figure out the actual
	// values in the cluster, due to multiple versions of CRDs installed.
	undefinedKeyword = "UNDEFINED"
```

**Review by shaneutt — APPROVED** [review]
Besides the comments already provided, looks good :+1: 

/approve

File: conformance/utils/suite/profiles.go
Comment (robscott) [reply conformance/utils/suite/profiles.go:71]: I think @dprotaso's concerns are actually the same as mine - how do we avoid exponential growth in terms of profiles here?

For example, you could imagine multiples of the following:

mesh+gateway * each unique combination of route types (by protocol, layer, or all) * core + extended

One option would be to start exclusively with profiles for core support of GA resources in v1.1 (Gateway, HTTPRoute, and GRPCRoute) - I think Gateway is inseparable from GatewayClass so it's probably fine to merge those together. Maybe to take that one step further, does that just become another way to express the `SupportHTTPRoute` supported feature?

To be clear, I'm very excited about the updated conformance reporting which I think is what the vast majority of this PR is about, I'm just worried that if we're not careful with the profiles terminology here, we'll end up with an unmaintainable number of profile names alongside an ever-expanding set of feature names that could become overly confusing when layered on top of each other.

Comment (shaneutt) [reply conformance/utils/suite/profiles.go:71]: As a reminder, the idea of profiles is that they're just convenient groupings of features, e.g.:

```
Individual Tests
|_
    SupportedFeatures
    |_
        ConformanceProfiles
```

The profiles are meant to cover the most common combinations of feature groupings, providing high level composability for implementations to subscribe to. This has been documented as the intent for over a year, and the currently expected profiles are common ones such as "HTTP", "Mesh", "L4", e.t.c.

I wonder if trying to change this now (especially as we were just about to go standard with it) would just be treating a symptom of a problem no where near the source of the problem, like how pain in us humans can radiate so we feel it somewhere it is not: perhaps Gateway APIs scope is way too large, and we're trying to cover too many things as a single project. So perhaps it is not "there's potential for too many conformance profiles" but instead "there's too many things we're trying to support as a single sub-project".

Comment (dprotaso) [reply conformance/utils/suite/profiles.go:71]: Hey @shaneutt thanks for the clarification. I'm good with this then. I also missed the fact that extended features need to be activated with the `SupportedFeatures` flag when running the test

Comment (shaneutt) [reply conformance/utils/suite/profiles.go:71]: Yes, gotcha. There are a couple other things to keep in mind here too:

1. Network Policy is (without any complaints thus far) using this as well and hoping to standardize on it once we do, so if we make any last minute large changes to the GEP this effects them as well.
2. Unlike the APIs or regular Go code, this doesn't have direct _user_ implications. So if we do find in time we need a revision because we start experiencing some real world problems, it's not as hard to make some backwards incompatible changes.

Since this PR is simply the last PR _implementing the existing GEP_, because we're 2 weeks before release, and given the above context I think we should move forward as is and judge this based on it's adherence to the current GEP, and if needed make grander changes to the GEP itself separately.

**Review by dprotaso — APPROVED** [review]

**Comment by youngnick** [comment]
LGTM once the merge conflict is fixed.

File: conformance/utils/suite/conformance.go
Comment (tao12345666333) [inline conformance/utils/suite/conformance.go:51]: :+1:  I recently discovered this problem.

**Review by tao12345666333 — APPROVED** [review]

File: conformance/conformance_test.go
Comment (tao12345666333) [inline conformance/conformance_test.go:104]: Is `*flags.RunTest` missing here?

Comment (mlavacca) [reply conformance/conformance_test.go:106]: added in https://github.com/kubernetes-sigs/gateway-api/pull/2868/commits/1fd59a4a8aa14033b7eb9270b2b7f14dfbd5ff09

File: conformance/utils/suite/suite.go
Comment (zirain) [inline conformance/utils/suite/suite.go:513]: @mlavacca do you think it's a good idea to have a knob(default to false) to skip this check? I just want to run conformance with previous version CRD and see what will happen.

Comment (zirain) [reply conformance/utils/suite/suite.go:513]: found `--allow-crds-mismatch`

**Pull request merged**
